### ISSUE TITLE
ExecutionId does not persist after execution has completed.

### ISSUE THREAD

**Issue by zleonov** [graphql-java/graphql-java#2992]
**Describe the bug**
When a new ExecutionInput is created via `ExecutionInput.Builder.build()` the `ExecutionInput.getExecutionId()` method initially returns a null unless it was explicitly set via `ExecutionInput.Builder.executionId(ExecutionId)`.

For example:
`final ExecutionInput input = ExecutionInput.newExecutionInput().query(query).build();`
`System.out.println(input.getExecutionId()); // prints null`

The `GraphQL.execute(ExecutionInput)` method will then set an executionId before commencing execution via the `GraphQL.ensureInputHasId(ExecutionInput)` method. This executionId can later be retreived from the DataFetchingEnvironment perhaps for logging purposes to identify each specific query.

However, that method actually creates a copy of the original ExecutionInput (with all the original values) save for the setting the newly generated executionId. Ostensibly this is because once the ExecutionInput is built, there is no ability to set the executionId. The only time it can be set is in the ExecutionInput.Builder itself:

See the last return statement from the method below (GraphQL.java - Line 544 in v19.2) :

```
   private ExecutionInput ensureInputHasId(ExecutionInput executionInput) {
        if (executionInput.getExecutionId() != null) {
            return executionInput;
        }
        String queryString = executionInput.getQuery();
        String operationName = executionInput.getOperationName();
        Object context = executionInput.getGraphQLContext();
        return executionInput.transform(builder -> builder.executionId(idProvider.provide(queryString, operationName, context)));
    }
```

The problem is after execution is complete the user does not have access to the cloned ExecutionInput or the executionId. There are many reasons why you'd want the executionId to persist after execution finishes. For example: you may want to log the errors which occurred via `ExecutionResult.getErrors()`, but without the executionId there is no way to link the errors to the original query or the prior log messages (which had access to the executionId from the DataFetchingEnvironment).

It also breaks the principle of least surprise. There is no reason without looking at the code, one would assume the ExecutionInput is cloned and the executionId is lost:

```
        final ExecutionInput input = ExecutionInput.newExecutionInput().query(query).build();
        final ExecutionResult result = graphql.execute(input);
        for (final GraphQLError error : result.getErrors())
            System.out.println("executionId: " + input.getExecutionId() + " error: " + error.getMessage());
            // this will print a null for the executionId, but that is not expected.
```

**Proposal**
The easiest fix would be to change the default behavior of ExecutionInput to allow the executionId to be set outside of the builder by creating a packaged scope `ExecutionInput.setExecutionId(ExecutionId)` method. That way if it has not been set there is no need to create a copy of the ExecutionInput in the `GraphQL.ensureInputHasId(ExecutionInput)` method. The downside of this is that ExecutionInput will lose its _immutability_ semantics, but to a very small degree if the method is package scoped. And this should not affect backward compatibility.

**Comment by bbakerman** [comment]
I would really really not like to change `ExecutionInput` so that its mutable.  That path leads to multi threaded made ness.

One though is to default the executionId say BUt that unwinds the idea that a `ExecutionIdProvider` can be used to provide values

I think given those constraints, the only option is for consumers to SET the executionId in themselves if they want it on the outside of the graphql call.

Can you explain why in your case you can jsut set `executionId` into the `ExecutionInput`?   I can gets maybe because its a framework issue but I wan to be sure

**Comment by zleonov** [comment]
### So first to answer your question: ###

I ended up doing exactly what you suggest. I put my own UUID as the executionId and that solved the problem.

But it took me a few weeks to get to the bottom of it. We have a multiuser system. I was debugging critical errors that were rarely occurring, but were fatal for the client request. I started as everyone would start, by adding log statements at the service layer and in the resolvers (we use graphql-java-tools) using the `executionId` from the `DataFetchingEnvironment`.

But that didn't immediately reveal the bug, so I decided to log all errors that occurred during the request. The natural place to do that is after the call returns via `ExecutionResult.getErrors()`. Otherwise, I'd be going down a rabbit hole since I didn't know where the error was coming from in the first place. I'd have to `try/catch` every single method in the code path.

So when I logged all the errors, imagine my surprise when the executionId was null. Since the critical errors in question were rarely occurring, and requests were coming in from multiple users, it wasn't really possible to correlate which requests caused the errors.

It didn't occur to me why the `executionId` was null after the call returned. Before looking at the graphql-java code, I thought something crazy was going on...like maybe we were clearing some ThreadLocal variable or something that was causing the `executionId` to be null.

### My thoughts... ###
I don't think the best choice is to mandate the consumer set his own `executionId` when building the `ExecutionInput`. Not because that's a hard mandate, in fact it's easy, and perhaps it should be the **_only_** way to set the executionId, that would make it clear. But because:

1. It makes the whole idea of using an `ExecutionIdProvider` and having GraphQL set the `executionId` for you, somewhat broken and inconsistent. That is to say, even given a custom `ExecutionIdProvider`, the the `executionId` will still be null after the call returns.

2. I think my use case is not rare, but is pretty common. The natural place to log uncaught errors is from the `ExeuctionResult` after the call returns. Especially if sanitizing them (perhaps) before returning a response to the client.

3. While I completely agree that it is best practice to make all objects as immutable as possible, especially in a concurrent environment, in this particular case I don't see how making a packeged scope `setExecutionId` method will cause some race condition. You know the code better, but as far as I can tell, the only place it will be set by graphql is in the `ensureInputHasId` method. And that would happen only by a single thread operating on a single `ExecutionInput` instance.

**Comment by bbakerman** [comment]
When I added `ExecutionId` into the mix it was because I figured that some system outside the graphql engine would be providing the unique id.

For example `B3-TraceIds` from something like ZipKin or Amazon `X-Amzn-Trace-Id` headers that get propagated by code into the mix.

These by their nature exist outside the engine.  

graphql-java aims to be opinionated about most things and it only really provides and execution id as a worst case situation if none has been set.   But it really intends other code to set it
